pkg/queryfrontend/codec.go

```diff
@@ -0,0 +1,307 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if result.End < result.Start {
+		return nil, errEndBeforeStart
+	}
+
+	result.Step, err = parseDurationMillis(r.FormValue("step"))
+	if err != nil {
+		return nil, err
+	}
+
+	if result.Step <= 0 {
+		return nil, errNegativeStep
+	}
+
+	// For safety, limit the number of returned points per timeseries.
+	// This is sufficient for 60s resolution for a week or 1h resolution for a year.
+	if (result.End-result.Start)/result.Step > 11000 {
+		return nil, errStepTooSmall
+	}
+
+	result.Dedup, err = parseEnableDedupParam(r.FormValue("dedup"))
+	if err != nil {
+		return nil, err
+	}
+
+	result.MaxSourceResolution, err = parseDownsamplingParamMillis(r.FormValue("max_source_resolution"))
+	if err != nil {
+		return nil, err
+	}
+
+	result.PartialResponse, err = parsePartialResponseParam(r.FormValue("partial_response"), c.partialResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(r.Form["replicaLabels[]"]) > 0 {
+		result.ReplicaLabels = r.Form["replicaLabels[]"]
+	}
+
+	result.StoreMatchers, err = parseStoreMatchersParam(r.Form["storeMatch[]"])
+	if err != nil {
+		return nil, err
+	}
+
+	result.Query = r.FormValue("query")
+	result.Path = r.URL.Path
+	return &result, nil
+}
+
+func (c codec) EncodeRequest(ctx context.Context, r queryrange.Request) (*http.Request, error) {
+	thanosReq, ok := r.(*ThanosRequest)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "invalid request format")
+	}
+	params := url.Values{
+		"start":            []string{encodeTime(thanosReq.Start)},
+		"end":              []string{encodeTime(thanosReq.End)},
+		"step":             []string{encodeDurationMillis(thanosReq.Step)},
+		"query":            []string{thanosReq.Query},
+		"dedup":            []string{strconv.FormatBool(thanosReq.Dedup)},
+		"partial_response": []string{strconv.FormatBool(thanosReq.PartialResponse)},
+		"replicaLabels[]":  thanosReq.ReplicaLabels,
+	}
+
+	// Add this param only if it is set. Set to 0 will impact
+	// auto-downsampling in the querier.
+	if thanosReq.MaxSourceResolution != 0 {
+		params["max_source_resolution"] = []string{encodeDurationMillis(thanosReq.MaxSourceResolution)}
+	}
+
+	if len(thanosReq.StoreMatchers) > 0 {
+		storeMatchers, err := matchersToStringSlice(thanosReq.StoreMatchers)
+		if err != nil {
+			return nil, httpgrpc.Errorf(http.StatusBadRequest, "invalid request format")
+		}
+		params["storeMatch[]"] = storeMatchers
+	}
+
+	u := &url.URL{
+		Path:     thanosReq.Path,
+		RawQuery: params.Encode(),
+	}
+	req := &http.Request{
+		Method:     "GET",
+		RequestURI: u.String(), // This is what the httpgrpc code looks at.
+		URL:        u,
+		Body:       http.NoBody,
+		Header:     http.Header{},
+	}
+
+	return req.WithContext(ctx), nil
+}
+
+func (c codec) DecodeResponse(ctx context.Context, r *http.Response, req queryrange.Request) (queryrange.Response, error) {
+	return c.prometheusCodec.DecodeResponse(ctx, r, req)
```

Comment:
We could in theory embed Prometheus codec so we don't need to even implement those methods (:

Reply:
Thanks, good catch! Let me update it.